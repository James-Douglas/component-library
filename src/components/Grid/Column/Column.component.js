import React from 'react';
import PropTypes from 'prop-types';
import StyledColumn from './Column.styles';

const Column = ({
  cols,
  sm,
  md,
  lg,
  xl,
  xxl,
  offset,
  offsetSm,
  offsetMd,
  offsetLg,
  offsetXl,
  offsetXxl,
  children,
  className,
  valign,
  halign,
}) => (
  <StyledColumn
    cols={cols}
    sm={sm}
    md={md}
    lg={lg}
    xl={xl}
    xxl={xxl}
    baseOffset={offset}
    offsetSm={offsetSm}
    offsetMd={offsetMd}
    offsetLg={offsetLg}
    offsetXl={offsetXl}
    offsetXxl={offsetXxl}
    valign={valign}
    halign={halign}
    className={`
        ${className}
      `}
  >
    {children}
  </StyledColumn>
);

Column.propTypes = {
  /**
   * Classes to be applied to the container element
   */
  className: PropTypes.string,
  /**
   * Width in columns (across all breakpoints where a specific width [sm, md, lg, xl, xxl] has not been given)
   */
  cols: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Width in columns at sm screen sizes.
   */
  sm: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Width in columns at md screen sizes.
   */
  md: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Width in columns at lg screen sizes.
   */
  lg: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Width in columns at xl screen sizes.
   */
  xl: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Width in columns at xxl screen sizes.
   */
  xxl: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Offset in columns (across all breakpoints where a specific offset [sm, md, lg, xl, xxl] has not been given)
   */
  offset: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Offset in columns at the sm screen size.
   */
  offsetSm: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Offset in columns at the md screen size.
   */
  offsetMd: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Offset in columns at the lg screen size.
   */
  offsetLg: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Offset in columns at the xl screen size.
   */
  offsetXl: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * Offset in columns at the xxl screen size.
   */
  offsetXxl: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
  ]),
  /**
   * The column contents.
   */
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  /**
   * The align-items property is related to CSS layout.
   */
  valign: PropTypes.oneOf([
    'flex-start',
    'flex-end',
    'center',
    'baseline',
    'stretch',
  ]),
  /**
   * The justify-content property is related to CSS layout.
   */
  halign: PropTypes.oneOf([
    'flex-start',
    'flex-end',
    'center',
    'space-between',
    'space-around',
    'initial',
    'inherit',
  ]),
};

Column.defaultProps = {
  className: '',
  cols: 0,
  sm: 0,
  md: 0,
  lg: 0,
  xl: 0,
  xxl: 0,
  offset: 0,
  offsetSm: 0,
  offsetMd: 0,
  offsetLg: 0,
  offsetXl: 0,
  offsetXxl: 0,
  children: [],
  valign: 'stretch',
  halign: 'initial',
};

export default Column;
