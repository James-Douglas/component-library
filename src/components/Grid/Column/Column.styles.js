import styled, { css } from 'styled-components';
import screens from '../../../../config/screens';

// for cols='sm' or cols='md' etc
const colDefault = css`
  flex-basis: 0%;
  flex-grow: 1;
  max-width: 100%;
`;
// for sm='auto' or md='auto' etc
const colAuto = css`
  flex: 0 0 auto;
  width: auto;
  max-width: 100%;
`;

const getPercentage = (value) => {
  const parsed = parseInt(value, 10);
  return `${(100 / 12) * parsed}%`;
};

const media = [
  { col: 'sm', minWidth: screens.sm },
  { col: 'md', minWidth: screens.md },
  { col: 'lg', minWidth: screens.lg },
  { col: 'xl', minWidth: screens.xl },
  { col: 'xxl', minWidth: screens.xxl },
];

const StyledColumn = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
  padding-left: ${({ theme }) => theme.spacing[16]};
  padding-right: ${({ theme }) => theme.spacing[16]};
  align-items: ${(props) => props.valign};
  justify-content: ${(props) => props.halign};

  ${({ cols, baseOffset }) => {
    let colStyles = '';
    if (parseInt(cols, 10) > 0) {
      colStyles += `
        flex: ${getPercentage(cols)};
        max-width: ${getPercentage(cols)};
      `;
    }

    if (baseOffset) {
      colStyles += `
        margin-left: ${getPercentage(baseOffset)};
      `;
    }

    media.forEach((oneMedia) => {
      if (cols === oneMedia.col) {
        colStyles += `@media (min-width: ${oneMedia.minWidth}) {
          ${colDefault}
        }`;
      }
    });

    return colStyles;
  }}

  @media (min-width: ${screens.xs}) {
    ${({ sm, offsetSm }) => {
    let colStyles = '';

    if (sm === 'auto') {
      return colAuto;
    }

    if (sm) {
      colStyles += `
          flex: ${getPercentage(sm)};
          max-width: ${getPercentage(sm)};
        `;
    }

    if (offsetSm) {
      colStyles += `margin-left: ${getPercentage(offsetSm)}`;
    }

    return colStyles;
  }}
  }

  @media (min-width: ${screens.sm}) {
    ${({ md, offsetMd }) => {
    let colStyles = '';

    if (md === 'auto') {
      return colAuto;
    }

    if (md) {
      colStyles += `
          flex: ${getPercentage(md)};
          max-width: ${getPercentage(md)};
        `;
    }

    if (offsetMd) {
      colStyles += `margin-left: ${getPercentage(offsetMd)}`;
    }

    return colStyles;
  }}
  }

  @media (min-width: ${screens.md}) {
    ${({ lg, offsetLg }) => {
    let colStyles = '';

    if (lg === 'auto') {
      return colAuto;
    }

    if (lg) {
      colStyles += `
          flex: ${getPercentage(lg)};
          max-width: ${getPercentage(lg)};
        `;
    }

    if (offsetLg) {
      colStyles += `margin-left: ${getPercentage(offsetLg)}`;
    }

    return colStyles;
  }}
  }

  @media (min-width: ${screens.lg}) {
    ${({ xl, offsetXl }) => {
    let colStyles = '';

    if (xl === 'auto') {
      return colAuto;
    }

    if (xl) {
      colStyles += `
          flex: ${getPercentage(xl)};
          max-width: ${getPercentage(xl)};
        `;
    }

    if (offsetXl) {
      colStyles += `margin-left: ${getPercentage(offsetXl)}`;
    }

    return colStyles;
  }}
  }

  @media (min-width: ${screens.xl}) {
    ${({ xxl, offsetXxl }) => {
    let colStyles = '';

    if (xxl === 'auto') {
      return colAuto;
    }

    if (xxl) {
      colStyles += `
          flex: ${getPercentage(xxl)};
          max-width: ${getPercentage(xxl)};
        `;
    }

    if (offsetXxl) {
      colStyles += `margin-left: ${getPercentage(offsetXxl)}`;
    }

    return colStyles;
  }}
  }

  ${(props) => !props.cols
    && !props.sm
    && !props.md
    && !props.lg
    && !props.xl
    && !props.xxl
    && css`
      -ms-flex-preferred-size: 0;
      flex-basis: 0%;
      -ms-flex-positive: 1;
      flex-grow: 1;
      max-width: 100%;
    `}
`;

export default StyledColumn;
