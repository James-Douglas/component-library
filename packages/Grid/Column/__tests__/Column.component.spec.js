import React from 'react';
import { render } from '@testing-library/react';
import 'jest-styled-components';
import Column from '../Column.component';

describe('Column', () => {
  it('renders correctly without props', () => {
    const { container } = render(<Column />);
    expect(container).toMatchSnapshot();
  });

  it('renders children', () => {
    const { container } = render(
      <Column>
        <div className="child-content">test</div>
      </Column>,
    );
    const childContent = container.querySelector('.child-content');
    expect(childContent).toBeInTheDocument();
  });

  it('allows additional classes to be passed through', () => {
    const { container } = render(<Column className="test-class" />);
    const test = container.getElementsByClassName('test-class');
    expect(test).toBeDefined();
  });

  it('renders a column in auto mode ', () => {
    const { container } = render(<Column />);
    const column = container.firstChild;
    expect(column).toHaveStyleRule('flex-grow', '1', {
      media: 'only screen and (min-width:0em)',
    });
  });

  it('renders an xs column with 6 cols width ', () => {
    const { container } = render(<Column xs={6} />);
    const column = container.firstChild;
    expect(column).toHaveStyleRule('flex-basis', '50%', {
      media: 'only screen and (min-width:0em)',
    });
  });

  it('renders an sm column with 6 cols width ', () => {
    const { container } = render(<Column sm={6} />);
    const column = container.firstChild;
    expect(column).toHaveStyleRule('flex-basis', '50%', {
      media: 'only screen and (min-width:48em)',
    });
  });

  it('renders an sm column with 6 cols width ', () => {
    const { container } = render(<Column md={6} />);
    const column = container.firstChild;
    expect(column).toHaveStyleRule('flex-basis', '50%', {
      media: 'only screen and (min-width:64em)',
    });
  });

  it('renders an sm column with 6 cols width ', () => {
    const { container } = render(<Column lg={6} />);
    const column = container.firstChild;
    expect(column).toHaveStyleRule('flex-basis', '50%', {
      media: 'only screen and (min-width:75em)',
    });
  });

  it('renders a column with cols at 6, lg at 2', () => {
    const { container } = render(<Column cols={6} lg={2} />);
    const column = container.firstChild;
    expect(column).toHaveStyleRule('flex-basis', '50%', {
      media: 'only screen and (min-width:0em)',
    });
    expect(column).toHaveStyleRule('flex-basis', '50%', {
      media: 'only screen and (min-width:48em)',
    });
    expect(column).toHaveStyleRule('flex-basis', '50%', {
      media: 'only screen and (min-width:64em)',
    });
    expect(column).toHaveStyleRule('flex-basis', '16.666666666666668%', {
      media: 'only screen and (min-width:75em)',
    });
  });

  it('renders a column with cols at 3, lg at 2, md at 1, with an offset of 1', () => {
    const { container } = render(<Column cols={3} md={1} lg={2} offset={1} />);
    const column = container.firstChild;
    expect(column).toHaveStyleRule('flex-basis', '25%', {
      media: 'only screen and (min-width:0em)',
    });
    expect(column).toHaveStyleRule('flex-basis', '25%', {
      media: 'only screen and (min-width:48em)',
    });
    expect(column).toHaveStyleRule('flex-basis', '8.333333333333334%', {
      media: 'only screen and (min-width:64em)',
    });
    expect(column).toHaveStyleRule('flex-basis', '16.666666666666668%', {
      media: 'only screen and (min-width:75em)',
    });
    expect(column).toHaveStyleRule('margin-left', '8.333333333333334%', {
      media: 'only screen and (min-width:0em)',
    });
    expect(column).toHaveStyleRule('margin-left', '8.333333333333334%', {
      media: 'only screen and (min-width:48em)',
    });
    expect(column).toHaveStyleRule('margin-left', '8.333333333333334%', {
      media: 'only screen and (min-width:64em)',
    });
    expect(column).toHaveStyleRule('margin-left', '8.333333333333334%', {
      media: 'only screen and (min-width:75em)',
    });
  });

  it('hides a column at a desired breakpoint', () => {
    const { container } = render(<Column cols={6} lg={false} />);
    const column = container.firstChild;
    expect(column).toHaveStyleRule('flex-basis', '50%', {
      media: 'only screen and (min-width:0em)',
    });
    expect(column).toHaveStyleRule('flex-basis', '50%', {
      media: 'only screen and (min-width:48em)',
    });
    expect(column).toHaveStyleRule('flex-basis', '50%', {
      media: 'only screen and (min-width:64em)',
    });
    expect(column).toHaveStyleRule('display', 'none', {
      media: 'only screen and (min-width:75em)',
    });
  });

  it('renders an offset', () => {
    const { container } = render(<Column xsOffset={1} />);
    const column = container.firstChild;
    expect(column).toHaveStyleRule('margin-left', '8.333333333333334%', {
      media: 'only screen and (min-width:0em)',
    });
  });

  it('renders a column with vertically aligned items', () => {
    const { container } = render(<Column valign="center" />);
    const column = container.firstChild;
    expect(column).toHaveStyleRule('align-items', 'center');
  });

  it('renders a column with horizontally aligned items', () => {
    const { container } = render(<Column halign="center" />);
    const column = container.firstChild;
    expect(column).toHaveStyleRule('justify-content', 'center');
  });
});
