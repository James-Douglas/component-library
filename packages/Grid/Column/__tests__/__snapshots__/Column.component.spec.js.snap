// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Column renders correctly without props 1`] = `
.c0 {
  box-sizing: border-box;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  padding-right: 0.5rem;
  padding-left: 0.5rem;
  box-sizing: border-box;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: initial;
  -webkit-justify-content: initial;
  -ms-flex-pack: initial;
  justify-content: initial;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

@media only screen and (min-width:0em) {
  .c0 {
    -webkit-box-flex: 1;
    -webkit-flex-grow: 1;
    -ms-flex-positive: 1;
    flex-grow: 1;
    -webkit-flex-basis: 0;
    -ms-flex-preferred-size: 0;
    flex-basis: 0;
    max-width: 100%;
    display: block;
  }
}

@media only screen and (min-width:48em) {
  .c0 {
    -webkit-box-flex: 1;
    -webkit-flex-grow: 1;
    -ms-flex-positive: 1;
    flex-grow: 1;
    -webkit-flex-basis: 0;
    -ms-flex-preferred-size: 0;
    flex-basis: 0;
    max-width: 100%;
    display: block;
  }
}

@media only screen and (min-width:64em) {
  .c0 {
    -webkit-box-flex: 1;
    -webkit-flex-grow: 1;
    -ms-flex-positive: 1;
    flex-grow: 1;
    -webkit-flex-basis: 0;
    -ms-flex-preferred-size: 0;
    flex-basis: 0;
    max-width: 100%;
    display: block;
  }
}

@media only screen and (min-width:75em) {
  .c0 {
    -webkit-box-flex: 1;
    -webkit-flex-grow: 1;
    -ms-flex-positive: 1;
    flex-grow: 1;
    -webkit-flex-basis: 0;
    -ms-flex-preferred-size: 0;
    flex-basis: 0;
    max-width: 100%;
    display: block;
  }
}

@media only screen and (min-width:0em) {
  .c0 {
    margin-left: 0%;
  }
}

@media only screen and (min-width:48em) {
  .c0 {
    margin-left: 0%;
  }
}

@media only screen and (min-width:64em) {
  .c0 {
    margin-left: 0%;
  }
}

@media only screen and (min-width:75em) {
  .c0 {
    margin-left: 0%;
  }
}

@media only screen and (min-width:0em) {
  .c0 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
  }
}

@media only screen and (min-width:48em) {
  .c0 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
  }
}

@media only screen and (min-width:64em) {
  .c0 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
  }
}

@media only screen and (min-width:75em) {
  .c0 {
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
  }
}

<div>
  <div
    class="c0 
        
      "
  />
</div>
`;
